src/app/rosetta/lib/block.ml

```diff
@@ -439,6 +439,314 @@ module Block_info = struct
+
+    let run (module Conn : Caqti_async.CONNECTION) id =
+      Conn.collect_list query id
+  end
+
+  module Public_keys = struct
+    (* TODO: Make this more efficient by batching or getting all in one query *)
+    let typ = Caqti_type.string
+
+    let query_one =
+      Caqti_request.find_opt Caqti_type.int typ
+        {| SELECT value FROM public_keys WHERE id = ? |}
+
+    let run (module Conn : Caqti_async.CONNECTION) id =
+      Conn.find_opt query_one id
+  end
+
+  let run (module Conn : Caqti_async.CONNECTION) input =
+    let module M = struct
+      include Deferred.Result
+
+      module List = struct
+        let map ~f =
+          List.fold ~init:(return []) ~f:(fun acc x ->
+              let open Let_syntax in
+              let%bind xs = acc in
+              let%map y = f x in
+              y :: xs )
+      end
+    end in
+    let open M.Let_syntax in
+    let%bind block_id, raw_block =
+      match%bind
+        Block.run (module Conn) input
+        |> Errors.Lift.sql ~context:"Finding block"
+      with
+      | None ->
+          M.fail (Errors.create `Block_missing)
+      | Some (block_id, raw_block) ->
+          M.return (block_id, raw_block)
+    in
+    let%bind parent_id =
+      match raw_block.parent_id with
+      | None ->
+          M.fail
+            (Errors.create ~context:"Parent block is null because genesis"
+               `Block_missing)
+      | Some id ->
+          M.return id
+    in
+    let%bind raw_parent_block =
+      match%bind
+        Block.run_by_id (module Conn) parent_id
+        |> Errors.Lift.sql ~context:"Finding parent block"
+      with
+      | None ->
+          M.fail (Errors.create ~context:"Parent block" `Block_missing)
+      | Some (_, raw_parent_block) ->
+          M.return raw_parent_block
+    in
+    let%bind raw_user_commands =
+      User_commands.run (module Conn) block_id
+      |> Errors.Lift.sql ~context:"Finding user commands within block"
+    in
+    let%bind raw_internal_commands =
+      Internal_commands.run (module Conn) block_id
+      |> Errors.Lift.sql ~context:"Finding internal commands within block"
+    in
+    let public_key_ids =
+      Int.Set.union_list
+        [ Int.Set.singleton raw_block.creator_id
+        ; List.fold raw_user_commands ~init:Int.Set.empty
+            ~f:(fun acc (_id, uc, _) ->
+              Int.Set.union acc
+                (Int.Set.of_list
+                   [ uc.Archive_lib.Processor.User_command.fee_payer_id
+                   ; uc.source_id
+                   ; uc.receiver_id ]) )
+        ; List.fold raw_internal_commands ~init:Int.Set.empty
+            ~f:(fun acc (_id, ic) ->
+              Int.Set.union acc
+                (Int.Set.singleton
+                   ic.Archive_lib.Processor.Internal_command.receiver_id) ) ]
+    in
+    let%bind public_keys =
+      (* TODO: N+1 problem *)
+      Int.Set.fold public_key_ids ~init:(M.return Int.Map.empty)
```

Comment:
We don't have `Int.Map.of_key_set` for the version of Core we use, but 
```ocaml
public_key_ids
|> Int.Set.to_sequence ~order:`Increasing
|> Async.Sequence.map ~f:(fun id -> let%bind pk = ... in (id, pk))
>>| Int.Map.of_increasing_sequence
```
should be `O(n)` rather than `O(n log(n))` when creating the map.